tls/s2n_ktls_parameters.h

```diff
@@ -24,21 +24,66 @@
  * - https://elixir.bootlin.com/linux/v6.3.8/A/ident/SOL_TCP
  */
 
-#define S2N_TLS_ULP_NAME      "tls"
-#define S2N_TLS_ULP_NAME_SIZE sizeof(S2N_TLS_ULP_NAME)
-
 #if defined(__linux__)
     #define S2N_KTLS_SUPPORTED true
 
     /* socket definitions */
     #define S2N_TCP_ULP 31 /* Attach a ULP to a TCP connection.  */
     #define S2N_SOL_TCP 6  /* TCP level */
+    #define S2N_SOL_TLS 282
+    #define S2N_TLS_TX  1 /* Set transmit parameters */
+    #define S2N_TLS_RX  2 /* Set receive parameters */
+
+    /* cipher definitions */
+    #define S2N_TLS_CIPHER_AES_GCM_128              51
+    #define S2N_TLS_CIPHER_AES_GCM_128_IV_SIZE      8
+    #define S2N_TLS_CIPHER_AES_GCM_128_KEY_SIZE     16
+    #define S2N_TLS_CIPHER_AES_GCM_128_SALT_SIZE    4
+    #define S2N_TLS_CIPHER_AES_GCM_128_TAG_SIZE     16
+    #define S2N_TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE 8
 
 #else
-    /* For unsupported platforms 0-init all values. */
-    #define S2N_KTLS_SUPPORTED false
+    /* For unsupported platforms 0-init (array of size 1) all values. */
+    #define S2N_KTLS_SUPPORTED                      false
 
     /* socket definitions */
-    #define S2N_TCP_ULP        0
-    #define S2N_SOL_TCP        0
+    #define S2N_TCP_ULP                             0
+    #define S2N_SOL_TCP                             0
+    #define S2N_SOL_TLS                             0
+    #define S2N_TLS_TX                              0
+    #define S2N_TLS_RX                              0
+
+    /* cipher definitions */
+    #define S2N_TLS_CIPHER_AES_GCM_128              0
+    #define S2N_TLS_CIPHER_AES_GCM_128_IV_SIZE      1
+    #define S2N_TLS_CIPHER_AES_GCM_128_KEY_SIZE     1
+    #define S2N_TLS_CIPHER_AES_GCM_128_SALT_SIZE    1
+    #define S2N_TLS_CIPHER_AES_GCM_128_TAG_SIZE     1
+    #define S2N_TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE 1
+
 #endif
+
+/* Common */
+#define S2N_TLS_ULP_NAME      "tls"
+#define S2N_TLS_ULP_NAME_SIZE sizeof(S2N_TLS_ULP_NAME)
+
+typedef unsigned short __u16;
+
```

Comment:
What is this typedef for? Why not just use uint16_t if that's what you need? Is `unsigned short` actually required to be uint16_t? I think it's just "usually" / minimally uint16_t.

Reply:
These are coming directly from linux source: https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/tls.h#L120

I am a bit torn about if we should be getting these or just defining them since we have to define the flag constants. Defining them makes out code simpler (less `ifdefs`), but you do have a point. I was relying on Linux backwards stability guarantees.

Lemme investigate

Reply:
Based on whats exposed in [uapi](https://github.com/torvalds/linux/blob/3c4d7559159bfe1e3b94df3a657b2cda3a34e218/include/uapi/linux/tls.h) when it was first exposed, (the [commit](https://github.com/torvalds/linux/commit/3c4d7559159bfe1e3b94df3a657b2cda3a34e218#diff-e23cbbcd372d08138b061222255e893a30d896c3875a33dd5d8bc32e21e0ca53)) we can eliminate a few flags/structs.

We would probably need to include the feature probe back in and match on `linux/tls.h` being present. However, as we can see these values have not changed since inception and they cant, since ppl rely on them and Linux will not break customers.

I think a good compromise might be to retrieve the `structs` but define the `flags`. Thoughts?

Reply:
I don't think the distinction is structs vs flags. Wherever possible, you need to use the value from the headers. You don't want to own implementation details you shouldn't own. We should only be defining values that we can't rely on the headers to provide.

Reply:
I have trimmed the values to those we cant rely on being available in the linux uapi. The downside of this is that there are more ifdefs in the codebase now.. but I think this is something we will have to live with.